### DISCUSSION TITLE
setBackgroundMessageHandler does not trigger on iOS

### DISCUSSION THREAD

**Discussion by hnqlv** [invertase/react-native-firebase#6408] [Q&A] [answered]
Hello fellows coders,

I'm trying to debug this for a while. The ` setBackgroundMessageHandler` never triggers on iOS. I can see the notification popping up when the app is on background but that's it.

```tsx
// src/index.js
messaging().setBackgroundMessageHandler(message => {
  console.log('[notification] ', message) // nothing happens
})
```

I've tried the following:

- Using `contentAvailable: true` doesn't display the notification
- Using `headers` within `apns` doesn't display the notification
- Removing `notifications` and using silently data does not have any effect
- Removing the `data: { payload: {} }` does not have any effect

The payload:

```diff
{
  tokens: [],
  notification: {
    title: 'Test',
    body: 'Oi oi',
  },
  data: { payload: '{"type": "call"}'},
  android: {},
  apns: {
    payload: {
      aps: {
       'content-available': 1,
       'mutable-content': 1,
        sound: {
          critical: true,
          name: 'default',
          volume: 1.0,
        },
      },
    },
  },
}
```

Using the v14.9.3

```package.json
"@react-native-firebase/app": "^14.9.3",
"@react-native-firebase/messaging": "^14.9.3",
```

Background modes is properly set on Xcode.

<img width="506" alt="Screenshot 2022-07-18 at 22 37 58" src="https://user-images.githubusercontent.com/123278/179645469-d4277db9-5ece-4bc4-bdde-cc58808b9b43.png">

What am I missing here?

Cheers!

**Comment by mikehardy** [comment]
If you include a notification block, it will not trigger anything in the background, it shows the notification but you get no control
When the user interacts with it you should see getInitialNotification contain the thing?

**Reply by hnqlv** [reply]
@mikehardy yes, I've remove the notification block but the problem still there. It doesn't really trigger anything inside the `setBackgroundMessageHandler` method.  The foreground messages are working.

**Reply by mikehardy** [reply]
Now you are probably in one of the (unfortunately - quite common) cases where you learn that data only messages have unreliable delivery in ios. **by design from Apple**

Make sure you have content-available set to 1
Make sure background refresh is on
Make sure the phone is fully charged up, maybe even plugged in!
Make sure you have granted notifications permissions to the app
Make sure the phone is not too hot (not kidding!)

Watch the Xcode console on the real device to see what the operating system thinks when the FCM arrives. Maybe Apple decides to wake up the app and deliver it, calling the handler, maybe it doesn't (and it should leave a message in the log as to why)

**Comment by hnqlv** [comment, marked as answer]
Hey @mikehardy,

I'm gonna answer my own question hoping it will help someone else digging the same hole.

In their own [documentation](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/pushing_background_updates_to_your_app) Apple highlights it:

> The system treats background notifications as low priority: you can use them to refresh your app’s content, but the system doesn’t guarantee their delivery

On top of that iOS may put your app in a "suspended state". You should see something like this in the device logs:

```
com.apple.sleep.WakeUpResultsNotificationRetry.activity:[{ 
  name: FileProtectionPolicy, 
  policyWeight: 0.010, 
  response: {
    Decision: Absolutely Must Not Proceed, 
    Score: 0.00, 
    Rationale: [...]
  }
}], FinalDecision: Absolutely Must Not Proceed}
```

Checking the devices logs again, I can see the push notification did arrive but because the app is in a "suspended state" the notification alone can't wake it up. Pay attention to the `wake: 0` param in the log:

```
[com.myapp] Request DUET delivers content-available push notification to application
[com.myapp] Received remote notification request 38DB-1D75 [ waking: 0, hasAlertContent: 1, hasSound: 1 hasBadge: 0 hasContentAvailable: 1 hasMutableContent: 1 pushType: Background]
```

For me use case because they are critical to the product, the only solution I found is to implement [PushKit](https://developer.apple.com/documentation/pushkit).

Cheers!

**Reply by mikehardy** [reply]
You may find that using a notification block in combination with notifee.app and an extension helper allow you to have control of presentation and reliable delivery, it's the solution many use

**Reply by anhnch** [reply]
It sounds like my case, when the iOS settings > General > Background App Refresh is OFF. You should turn it on. If the device is in the Low Power Mode, it's turned OFF automatically. Notification is like just a little change makes it not working.

**Reply by hnqlv** [reply]
@anhnch yes indeed. Because I’m working on a voice call app I found the voip push the best alternative. I did the notification service extension as well and it does a good job overall.

**Reply by prijai07** [reply]
@mikehardy, can you provide some references on how to use notifee with notification block? 

I'm trying to update the app badge count using notifee, but facing the same issue of setBackgroundMessageHandler not triggering.

My index.js :

import {AppRegistry} from 'react-native';
import App from './App';
import {name as appName} from './app.json';
import messaging from '@react-native-firebase/messaging';
import notifee, {EventType} from '@notifee/react-native';

messaging().setBackgroundMessageHandler(async remoteMessage => {
  console.debug('Received background notification: ', remoteMessage);
  const {notification, data} = remoteMessage;

   if (Platform === 'ios') {
     // displays background notifications
     notifee.displayNotification({
       title: notification.title,
       body: notification.body,
       data,
     });
   }

  notifee.incrementBadgeCount();
  return Promise.resolve();
});

notifee.onBackgroundEvent(async ({type, detail}) => {
  const {notification, pressAction} = detail;
  if (type === EventType.PRESS) {
    await notifee.decrementBadgeCount();
    await notifee.cancelNotification(notification.id);
  }
});

AppRegistry.registerComponent(appName, () => App);

**Reply by frankA10019** [reply]
@prijai07 I am facing the same issue, any solution?

**Reply by ROG3R-DEV** [reply]
@prijai07 @frankA10019 I am facing the same issue, any solution?

**Reply by zahoormohammed** [reply]
i need help, i am facing the same issue

**Comment by AbaidullahRajpoot** [comment]
**----------------------------Am also facing this issue kindly help me-------------------------**

**---------------Here is my App.js file code-----------------------**

/**
 * @format
 */

import { AppRegistry } from 'react-native';
import App from './App';
import { name as appName } from './app.json';
import messaging from '@react-native-firebase/messaging';
import { Text, TextInput } from 'react-native';

//----------------Request  Permissions for Notifications ----------------\\

async function requestUserPermission() {
  const authStatus = await messaging().requestPermission();
  const enabled =
    authStatus === messaging.AuthorizationStatus.AUTHORIZED ||
    authStatus === messaging.AuthorizationStatus.PROVISIONAL;
  if (enabled) {
    console.log('Authorization status:', authStatus);
  }
}
requestUserPermission();

//----------------Call when App is in kill or  background state------//

messaging().setBackgroundMessageHandler(async remoteMessage => {
  console.log('Message handled in the background!', remoteMessage);
});

function HeadlessCheck({ isHeadless }) {
  if (isHeadless) {
    // App has been launched in the background by iOS, ignore
    return null;
  }

  // Render the app component on foreground launch
  return <App />;
}


AppRegistry.registerComponent(appName, () => HeadlessCheck);

//---------------disable font scaling IOS----------------------

if (Text.defaultProps == null) {
  Text.defaultProps = {};
  Text.defaultProps.allowFontScaling = false;
  TextInput.defaultProps = TextInput.defaultProps || {};
  TextInput.defaultProps.allowFontScaling = false;
}

**---------------Here is my AppDelegate.mm file code----------------------**

#import "RNFBMessagingModule.h"
#import <Firebase.h>
#import "AppDelegate.h"
#import <React/RCTBundleURLProvider.h>
//Facebook Setting
#import <AuthenticationServices/AuthenticationServices.h>
#import <SafariServices/SafariServices.h>
#import <FBSDKCoreKit/FBSDKCoreKit-swift.h>
#import <GoogleMaps/GoogleMaps.h>

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  
  self.moduleName = @"ShoptaPlace";
  [FIRApp configure];
  self.initialProps = [RNFBMessagingModule addCustomPropsToUserProps:nil withLaunchOptions:launchOptions];
  [[FBSDKApplicationDelegate sharedInstance] application:application
                       didFinishLaunchingWithOptions:launchOptions];
  [GMSServices provideAPIKey:@""]; // add this line using the api key obtained from Google Console
  // You can add your custom initial props in the dictionary below.
  // They will be passed down to the ViewController used by React Native.
  self.initialProps = @{};

  return [super application:application didFinishLaunchingWithOptions:launchOptions];
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index"];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

/// This method controls whether the `concurrentRoot`feature of React18 is turned on or off.
///
/// @see: https://reactjs.org/blog/2022/03/29/react-v18.html
/// @note: This requires to be rendering on Fabric (i.e. on the New Architecture).
/// @return: `true` if the `concurrentRoot` feature is enabled. Otherwise, it returns `false`.
- (BOOL)concurrentRootEnabled
{
  return true;
}

@end


**-----------------Packages I am using--------------------** 

 "react-native": "0.71.4",
 "@notifee/react-native": "^7.8.2"
 "@react-native-firebase/app": "^18.5.0",
 "@react-native-firebase/auth": "^18.5.0",
 "@react-native-firebase/dynamic-links": "^18.5.0",
 "@react-native-firebase/messaging": "^18.5.0",

**Comment by ducsilva** [comment]
Is there any way? It doesn't activate so I can't update the Badge

**Comment by zahoormohammed** [comment]
I am facing the same issue, 
messaging().setBackgroundMessageHandler(async (remoteMessage) => 

never triggers in background on IOS. but the same block is working on android. IOS is receiving the notification but setBackgroundMessageHandler is not triggering. i need help
